File: Codeforces/1144/f.py
import os 
from collections import deque, defaultdict

def isbipartite(col, g):
    col[1] = 0
    q = deque()
    q.append(1)
    bipartite = True
    while q:
        v = q.popleft()
        for i in range(0, len(g[v])):
            if col[g[v][i]] == -1:
                col[g[v][i]] = col[v] ^ 1
                q.append(g[v][i])
            else:
                bipartite = bipartite & (col[g[v][i]] != col[v])

    return bipartite

def main():
    N = 200100
    g = defaultdict(list)
    col = [-1]*N
    src = []
    dest = []
    n, m = map(int, input().split(" "))

    for i in range(0, m):
        s, d = map(int, input().split(" "))
        g[s].append(d)
        g[d].append(s)
        src.append(s)
        dest.append(d)
    
    edges = list(zip(src, dest))

    if not isbipartite(col, g):
        print("NO")
        exit(0)

    ans = ""

    for i in range(0, m):
        if col[edges[i][0]] == 0:
            ans = ans + "1"
        else:
            ans = ans + "0"

    print("YES")
    print(ans)

if __name__ == "__main__":
    main()

File: Codeforces/1144/f.cpp
#include<bits/stdc++.h>

using namespace std;

typedef long long ll;

#define N 200100

vector<ll> g[N];
vector<ll> col(N, -1);
vector<pair<ll,ll> > edges;
bool isb = true;
int n, m;

bool isbipartite(){
    queue<int> q;
    q.push(1);
    col[1] = 0;
    while(!q.empty()){
        int v = q.front();
        q.pop();
        for(auto i : g[v]){
            if(col[i] == -1){
                col[i] = col[v] ^ 1;
                q.push(i);
            }else{
                isb = isb & (col[i] != col[v]);
            }
        }
    }   
}

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    cin>>n>>m;

    //cleanup
    edges.clear();
    
    for(int i=1 ; i<=m ; i++){
        int s, d;
        cin>>s>>d;
        g[s].push_back(d);
        g[d].push_back(s);
        edges.push_back({s, d});
    }

    isbipartite();

    if(!isb){
        cout<<"NO"<<endl;
        exit(0);
    }

    string ans="";

    for(int i=0 ; i<edges.size() ; i++){
        if(col[edges[i].first] == 0){
            ans += "0";
        }else{
            ans += "1";
        }
    }
    
    cout<<"YES"<<endl;
    cout<<ans<<endl;

    return 0;
}